DOCS-2095: First draft of controls package docs update

martha-johnston opened the pull request:
# Description

- Update the controls package doc page to include the helper functions.
- move one example from encoded motor to sensor controlled base as the video was showing sensor controlled base and not motor... sorry I didn't catch this in previous reviews!

Not quite done yet but this is the majority of the work. I probably need some help cleaning things up and with the linking.

martha-johnston commented on docs/components/base/sensor-controlled.md:
```diff
@@ -80,11 +80,13 @@ The following attributes are available for `sensor-controlled` bases:
 ### Spin
 
-Spin implements a basic form of feedback control that, for a spin of less than 360 degrees, monitors how far the base has spun and stops once it has reached the goal position.
+When the `control_parameters` attribute is set, `Spin` implements a form of feedback control that polls the provided movement sensor and corrects any error between the desired angular velocity and the actual angular velocity using a PID control loop. `Spin` also monitors the angular distance traveled and stops the base when the goal angle is reached.
```
small update here because spin uses a control loop now!

martha-johnston commented on docs/components/base/sensor-controlled.md:
```diff
@@ -94,3 +96,11 @@ The following base control API methods are available on a `sensor-controlled` ba
+
+The position, orientation, and linear and angular velocity of the rover changing as it moves, as measured by a [movement sensor](/components/movement-sensor/):
+
+{{<gif webm_src="/components/encoded-motor/controls_change.webm" mp4_src="/components/encoded-motor/controls_change.mp4" alt="The control tab of a movement sensor on a base with encoded motors as it turns">}}
```
as mentioned in the description, this video used to be in encoded motor but it is an example of the sensor controlled base so I moved it

martha-johnston commented on docs/components/motor/gpio/encoded-motor.md:
```diff
@@ -148,7 +148,7 @@ In addition to the [attributes for a non-encoded motor](/components/motor/gpio/)
 | `ticks_per_rotation` | int | **Required** | Number of ticks in a full rotation of the encoder and motor shaft. |
 | `ramp_rate` | float | Optional | Rate to increase the motor's input voltage (power supply) per second when increasing the speed the motor rotates (RPM). <br> Range = (`0.0`, `1.0`] <br> Default: `0.2` |
-| `control_parameters` | object | Optional | A JSON object containing the coefficients for the proportional, integral, and derivative terms. If you want these values to be auto-tuned, you can set all values to 0: `{ "p": 0, "i": 0, "d": 0 }`, and `viam-server` will auto-tune and log the calculated values. Tuning takes several seconds and spins the motors. Copy the values from the logs and add the ...
+| `control_parameters` | object | Optional | A JSON object containing the coefficients for the proportional, integral, and derivative terms. If you want these values to be auto-tuned, you can set all values to 0: `{ "p": 0, "i": 0, "d": 0 }`, and `viam-server` will auto-tune and log the calculated values. Tuning takes several seconds and spins the motors. Copy the values from the logs and add the ...
```
a little bit of a wording change just because I think sensor is a bit confusing for motors

martha-johnston commented on docs/components/base/sensor-controlled.md:
```diff
@@ -80,11 +80,13 @@ The following attributes are available for `sensor-controlled` bases:
 
 ### SetVelocity
 
+If you want to control your base by specifying a desired velocity in terms of distance/time, for example 5 m/s, and you have a movement sensor that measures the actual velocity, you can make use of the sensor to adjust the velocity. Alternatively, if your base has position reporting motors, you can utilize the [wheeled odometry](/components/movement-sensor/wheeled-odometry/) movement sensor to ge ...
```
nvm I think I figured it out

martha-johnston commented on static/include/components/base-sensor.md:
```diff
@@ -1,12 +1,6 @@
-If you want to control your base by specifying a desired velocity in terms of distance/time, for example 5 m/s, and you have a sensor, for example a wheel encoder, that measures the actual velocity, you can make use of the sensor to adjust the velocity.
+If you want to control your base by specifying a desired velocity in terms of distance/time, for example 5 m/s, and you have a movement sensor that measures the actual velocity, you can make use of the sensor to adjust the velocity. Alternatively, if your base has position reporting motors, you can utilize the [wheeled odometry](../../../docs/components/movement-sensor/wheeled-odometry) movement  ...
```
thought I figured it out... but I did not. this link isn't passing the checks, let me know if you know of a better way to do this.

npentrel replied:
```suggestion
If you want to control your base by specifying a desired velocity in terms of distance/time, for example 5 m/s, and you have a movement sensor that measures the actual velocity, you can make use of the sensor to adjust the velocity.
Alternatively, if your base has position reporting motors, you can use the [wheeled odometry](/docs/components/movement-sensor/wheeled-odometry/) movement sensor to get an estimate of the necessary velocities.
```

martha-johnston commented on static/include/components/base-sensor.md:
```diff
@@ -1,12 +1,6 @@
-- A _derivative_ term that is the rate of change of the error
-
-By tuning the coefficients on each of these terms, you can adjust how your base converges towards the target value, how quickly the system reaches the target value, and how much the system overshoots when approaching the target value.
+Setting the `control_parameters` attribute sets up a PID control loop. Setting the `control_parameters` will automatically set up the required PID loop for a sensor controlled base. For more information on PID or to set up a more complex control loop, see the [controls package](../../../docs/internals/controls-package)
```
this link also isn't passing checks

npentrel replied:
```suggestion
Setting the `control_parameters` attribute sets up a PID control loop.
Setting the `control_parameters` will automatically set up the required PID loop for a sensor controlled base.
For more information on PID or to set up a more complex control loop, see the [controls package](/internals/controls-package/)
```

martha-johnston commented on static/include/components/motor-sensor.md:
```diff
@@ -2,15 +2,9 @@ If you want to control your motor by specifying the distance and velocity in ter
-- A _derivative_ term that is the rate of change of the error
-
-By tuning the coefficients on each of these terms, you can adjust how your motor converges towards the target values, how quickly the system reaches the target values, and how much the system overshoots when approaching the target values.
+The second requires setting the `control_parameters` attribute, which sets up a PID control loop to adjust the distance and velocity of the motor. Setting the `control_parameters` will automatically set up the required PID loop for an encoded motor. For more information on PID or to set up a more complex control loop, see the [controls package](../../../docs/internals/controls-package)
```
also not passing checks

npentrel replied:
it's complaining about the missing slash and about you having docs in the front I think?
```suggestion
The second requires setting the `control_parameters` attribute, which sets up a PID control loop to adjust the distance and velocity of the motor.
Setting the `control_parameters` will automatically set up the required PID loop for an encoded motor.
For more information on PID or to set up a more complex control loop, see the [controls package](/internals/controls-package/)
```

martha-johnston commented on docs/internals/controls-package.md:
```diff
@@ -9,15 +9,275 @@ description: "The Control Package implements feedback control on an endpoint."
+
+### BlockConfig
+
+`BlockConfig` is struct of the configuration for an individual control block. You will have to build individual BlockConfigs if you utilize the `UseCustomConfig` option. Each block type requires different attributes, which are outlined in the [Control Blocks](/internals/controls-package/#controlblocks) section.
```
this and the link on line 265 also aren't passing checks... but I think that probably makes sense because there's a space in the `Control Blocks` title... how do I account for that in the hash?

npentrel commented on docs/internals/controls-package.md:
```diff
@@ -9,20 +9,280 @@ description: "The Control Package implements feedback control on an endpoint."
 
 ## Introduction
 
-The control package implements feedback control on an endpoint (plant).
```
I don't know why this said plant haha

npentrel commented on docs/internals/controls-package.md:
```diff
@@ -9,20 +9,280 @@ description: "The Control Package implements feedback control on an endpoint."
+
+### BlockConfig
+
+`BlockConfig` is struct of the configuration for an individual control block. You will have to build individual BlockConfigs if you utilize the `UseCustomConfig` option. Each block type requires different attributes, which are outlined in the [Control Blocks](/internals/controls-package/#controlblocks) section.
```
You can see it here but essentially it's a dash for the space :) https://docs-test.viam.dev/2739/internals/controls-package/#control-blocks

```suggestion
`BlockConfig` is a struct for the configuration of an individual control block.
You have to build individual BlockConfigs if you use the `UseCustomConfig` option.
Each block type requires different attributes, which are outlined in the [Control Blocks](/internals/controls-package/#control-blocks) section.
```

npentrel commented on docs/internals/controls-package.md:
```diff
@@ -136,8 +396,8 @@ The Constant block outputs a constant signal. S_out = Cte
-For now, only DC motors with an encoder are supported as an endpoint in the control package.
-The _motor_name_ attribute is unused for now, and one should pass a Controllable interface when creating the loop.
+For now, only DC motors with an encoder, and a base with a linear and angular movement sensor are supported as an endpoint in the control package.
+The _motor_name_ attribute should be changed to _base_name_ when using a base.
```
We try to speak to the user as much as possible because otherwise they may think we will change it. This makes it clear they need to perform the action
```suggestion
You should change the _motor_name_ attribute to _base_name_ when using a base.
```

npentrel commented on docs/internals/controls-package.md:
```diff
@@ -172,7 +432,17 @@ There are two main approaches that one can use:
-Several methods to calculate Kp, Ki and Kd are implemented.
+Several methods to calculate Kp, Ki and Kd are implemented:
+
+- ziegerNicholsPI
```
well these seem like very random names. Do we want to explain any of this further?

martha-johnston replied:
they are tuning different tuning methods for a PID loop. I don't think we need to add more info as it is very technical and not really viam related. maybe a note that you can google them if you want more info but idk if that's necessary

npentrel replied:
Ok let's leave as is

npentrel reviewed (changes requested):
This is a great start - you're a good writer. I think I am confused what the controls package is supposed to be used for. Am I supposed to read this page if I want to implement my own feedback control? Can you give me an example of what I can do with this? I am a bit confused.

martha-johnston commented on docs/internals/controls-package.md:
```diff
@@ -9,20 +9,280 @@ description: "The Control Package implements feedback control on an endpoint."
+**Parameters:**
+
+- `ctx` [(Context)](https://pkg.go.dev/context): A context carries a deadline, a cancellation signal, and other values across API boundaries.
+- `name` [(string)](https://pkg.go.dev/builtin#string): The desired name of the trapezoidal block.
```
it's the name for the block that's being created by the function. so it could be anything. to replace an old block the name must be the same but if creating a new one completely it can be named anything

npentrel replied:
Ok but whatever I specify here will be used and not changed by other code? Then we want to remove "desired" because it can imply that you're only specifying a preference

martha-johnston replied:
yeah we don't want it to be changed it should be the same every time... that make sense

martha-johnston commented on docs/internals/controls-package.md:
```diff
@@ -9,20 +9,284 @@ description: "The Control Package implements feedback control on an endpoint."
+[`StartControlLoop`](/internals/controls-package/#startcontrolloop) | Starts the control loop in a background thread.
+[`CreateConstantBlock`](/internals/controls-package/#createconstantblock) | Creates a control block of type `constant`, all control loops need at least one constant block representing the set point.
+[`UpdateConstantBlock`](/internals/controls-package/#updateconstantblock) | Updates the value of a constant block to the new set point.
+[`CreateTrapzBlock`](/internals/controls-package/#createtrapzblock) | Creates a control block of type `trapezoidalVelocityProfile`, control loops, such as for motors, that control position need a trapezoidal velocity profile block.
```
```suggestion
[`CreateTrapzBlock`](/internals/controls-package/#createtrapzblock) | Creates a control block of type `trapezoidalVelocityProfile`. Control loops that control position (for example, encoded motors), need a trapezoidal velocity profile block.
```
does this make more sense

martha-johnston commented on docs/internals/controls-package.md:
````diff
@@ -9,20 +9,280 @@ description: "The Control Package implements feedback control on an endpoint."
+}
+```
+
+### Options
````
yep this is linked above where options is first mentioned

npentrel replied:
Yep sorry i missed that

martha-johnston commented on docs/internals/controls-package.md:
```diff
@@ -9,20 +9,280 @@ description: "The Control Package implements feedback control on an endpoint."
+
+- `pidVals` [([]PIDConfig)](/internals/controls-package/#pidconfig): The P, I, and D values for the control loop, if all are zero the loop will be auto-tuned and log the calculated PID values.
+- `componentName` [(string)](https://pkg.go.dev/builtin#string): The name of the component that the created PID loop will control.
+- `options` [(Options)](/internals/controls-package/#options): All the desired optional parameters to customize the control loop.
```
yes starting line 237

npentrel replied:
Sorry I missed the that the links were already there

martha-johnston commented on docs/internals/controls-package.md:
```diff
@@ -9,20 +9,280 @@ description: "The Control Package implements feedback control on an endpoint."
+- `pidVals` [([]PIDConfig)](/internals/controls-package/#pidconfig): The P, I, and D values for the control loop, if all are zero the loop will be auto-tuned and log the calculated PID values.
+- `componentName` [(string)](https://pkg.go.dev/builtin#string): The name of the component that the created PID loop will control.
+- `options` [(Options)](/internals/controls-package/#options): All the desired optional parameters to customize the control loop.
+- `c` [(Controllable)](/internals/controls-package/#controllable): An interface that contains the necessary functions to move the controlled component.
```
yes starting line 221

martha-johnston commented on docs/internals/controls-package.md:
````diff
@@ -9,20 +9,280 @@ description: "The Control Package implements feedback control on an endpoint."
+}
+```
+
+### PIDConfig
````
you don't really need it but if you want to really customize things and do it manually you would need it

martha-johnston commented:
> This is a great start - you're a good writer. I think I am confused what the controls package is supposed to be used for. Am I supposed to read this page if I want to implement my own feedback control? Can you give me an example of what I can do with this? I am a bit confused.

The main reason would be if you wanted to set up a control loop outside of the standard way that rdk sets it up for the two builtin components (encoded motor and sensor controlled base)

martha-johnston commented on docs/internals/controls-package.md:
```diff
@@ -9,20 +9,280 @@ description: "The Control Package implements feedback control on an endpoint."
+[`StartControlLoop`](/internals/controls-package/#startcontrolloop) | Starts the control loop in a background thread.
+[`CreateConstantBlock`](/internals/controls-package/#createconstantblock) | Creates a control block of type `constant`, all control loops need at least one constant block representing the set point.
+[`UpdateConstantBlock`](/internals/controls-package/#updateconstantblock) | Updates the value of a constant block to the new set point.
+[`CreateTrapzBlock`](/internals/controls-package/#createtrapzblock) | Creates a control block of type `trapezoidalVelocityProfile`, control loops, such as for motors, that control position need a trapezoidal velocity profile block.
```
```suggestion
[`CreateTrapzBlock`](/internals/controls-package/#createtrapzblock) | Creates a control block of type `trapezoidalVelocityProfile`. Control loops that control position (for example, encoded motors), need a trapezoidal velocity profile block.
```
does this make more sense

npentrel replied:
yes I think maybe add a "for"
```suggestion
[`CreateTrapzBlock`](/internals/controls-package/#createtrapzblock) | Creates a control block of type `trapezoidalVelocityProfile`. Control loops that control position (for example, control loops for encoded motors), need a trapezoidal velocity profile block.
```

martha-johnston commented:
@npentrel just added some more examples. I can remove them if they're too much, but I'm hoping it adds some clarity for the use cases

npentrel commented on docs/internals/controls-package.md:
```diff
@@ -9,20 +9,284 @@ description: "The Control Package implements feedback control on an endpoint."
+
+- `pidVals` [([]PIDConfig)](/internals/controls-package/#pidconfig): The P, I, and D values for the control loop, if all are zero the loop will auto-tune and log the calculated PID values.
+- `componentName` [(string)](https://pkg.go.dev/builtin#string): The name of the component that the PID loop controls.
+- `options` [(Options)](/internals/controls-package/#options): All the desired optional parameters to customize the control loop.
```
```suggestion
- `options` [(Options)](/internals/controls-package/#options): All the desired [optional parameters](#options) to customize the control loop.
```

martha-johnston replied:
is it necessary to link twice in the same statement? if this is the norm that's fine I don't feel strongly about it but I'm just wondering if we need the type "Options" and the phrase "optional parameters" to link to the same spot

npentrel reviewed (commented):
Addressed comments but haven't yet written/adjusted the intro to include the info you provided on what someone might use this for. I'll do that later (currently on mobile)

martha-johnston commented:
> Addressed comments but haven't yet written/adjusted the intro to include the info you provided on what someone might use this for. I'll do that later (currently on mobile)

Good point, I can update the intro!

martha-johnston commented on docs/internals/controls-package.md:
```diff
@@ -9,20 +9,286 @@ description: "The Control Package implements feedback control on an endpoint."
 ## Introduction
 
-The control package implements feedback control on an endpoint (plant).
+The easiest way to utilize the controls package is by configuring the `control_parameters` attribute for an [encoded motor](/docs/components/motor/gpio/encoded-motor/) and/or a [sensor-controlled base](/docs/components/base/sensor-controlled/). However, if you want to change or customize the control loops on these components, or you want to add a control loop to a different component, you can use ...
```
not the most info here, let me know if you have any suggestions on expanding on this

martha-johnston commented on docs/internals/controls-package.md:
```diff
@@ -4,25 +4,291 @@ linkTitle: "Control Package"
 
-## Introduction
+[Encoded motors](/components/motor/gpio/encoded-motor/) and [sensor controlled bases](/components/base/sensor-controlled/) use a control loop that is implemented by `viam-server`.
+You can configure the `control_parameters` attribute forn both components to adjust the control loop.
```
```suggestion
You can configure the `control_parameters` attribute for both components to adjust the control loop.
```

npentrel reviewed (approved):
Thank you for your patience on my review - R2D2 took a bit more work than I expected. I've adjusted the intro slightly and am happy for you to merge. Thank you very much for this detailed page!

martha-johnston commented:
@npentrel no worries! I'm happy with the current state of everything so yeah feel free to make any other adjustments you want, including the couple of suggestions you left, and merge whenever you're happy with it.
